src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java

```diff
@@ -0,0 +1,200 @@
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.ControlledAiNotebookInstanceResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.notebooks.v1.model.ContainerImage;
+import com.google.api.services.notebooks.v1.model.Instance;
+import com.google.api.services.notebooks.v1.model.Operation;
+import com.google.api.services.notebooks.v1.model.VmImage;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * A step for creating the AI Platform notebook instance in the Google cloud.
+ *
+ * <p>Undo deletes the created notebook instance.
+ */
+public class CreateAiNotebookInstanceStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(CreateAiNotebookInstanceStep.class);
+  private final CrlService crlService;
+  private final ControlledAiNotebookInstanceResource resource;
+  private final WorkspaceService workspaceService;
+
+  public CreateAiNotebookInstanceStep(
+      CrlService crlService,
+      ControlledAiNotebookInstanceResource resource,
+      WorkspaceService workspaceService) {
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    String projectId = workspaceService.getRequiredGcpProject(resource.getWorkspaceId());
+    InstanceName instanceName = createInstanceName(projectId);
+    Instance instance = createInstanceModel(flightContext, projectId);
+
+    AIPlatformNotebooksCow notebooks = crlService.getAIPlatformNotebooksCow();
+    try {
+      OperationCow<Operation> creationOperation;
+      try {
+        creationOperation =
+            notebooks
+                .operations()
+                .operationCow(notebooks.instances().create(instanceName, instance).execute());
+      } catch (GoogleJsonResponseException e) {
+        // If the instance already exists, this step must have already run successfully. Otherwise
+        // retry.
+        if (HttpStatus.CONFLICT.value() == e.getStatusCode()) {
+          logger.debug("Notebook instance {} already created.", instanceName.formatName());
+          return StepResult.getStepResultSuccess();
+        }
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+      }
+
+      creationOperation =
+          OperationUtils.pollUntilComplete(
+              creationOperation, Duration.ofSeconds(20), Duration.ofMinutes(12));
+      if (creationOperation.getOperation().getError() != null) {
+        throw new RetryException(
```

Comment:
Yes. I don't know if there's a preferred style here. Undo's can't throw a RetryException, so maybe it's better to always return a StepResult?
This doesn't have an explicit exception, so without throwing a RetryException I wasn't sure how to encode any message for Stairway debugging what happened. A `StepResult` doesn't have its own error message field, other than an `Exception`. Rather than invent an exception, I went with throwing a new RetryException with my own message.
Happy to do something else here.

Reply:
No need to change it. It just struck me odd looking at it. I don't think I've seen a case before where requesting Stairway retry is not in response to an exception. If this turns up more frequently, it might be worth adding a StepResult that takes a String instead of Exception